Collective Effects Config in Dashboard

proy30 opened the pull request:
PR adds in Space Charge and CSR configuration for users.

TODO:
-  [X] Space Charge Defaults
      - poisson_solver = "fft" (though, in ImpactX documentation default is "multigrid".)
      - max_level = 0
      - particle_shape = 2
      - n_cell = 1 (for x,y,z)
      -  blocking_factor (x,y,z) - 32
-  [X] Space Charge Validation
      - [X] **n_cell** - value_type = 'int' & >=1, must be a factor of `blocking_factor`
      - [X] **prob_relative** - if poisson_solver = "fft", index[0] >=3
                                            if possion_solver = "multigrid", index[0] >1
                                           else index[1-(n-1)] >1
    - [X] **blocking_factor (x,y,z)** - Greater than 0
-  [X] CSR Defaults
      - csr_bins = 150
-  [X] CSR Validation
     - csr_bins='int', positive number
-  [X] Add multigrid-specific numerical options (no validation as of now)
    - MLMG Relative Tolerance
    -  MLMG Absolute Tolerance
    - MLMG Max Iterations
    - MLMG Verbosity 
- [x] Update export_template
- [x] Update run_simulation status function

Resolves #665

proy30 committed:
Added initial space charge UI with basic compatibility with simulation

proy30 committed:
Added initial space charge UI with basic compatibility with simulation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Connected max_level and poisson_solver to simulation

proy30 committed:
prob_relative simulation functionality added

proy30 committed:
prob_relative simulation functionality added

proy30 committed:
Update code structuring

proy30 committed:
Update code structuring

proy30 committed:
rename variables

proy30 committed:
rename variables

proy30 committed:
Moved input cards

proy30 committed:
Moved input cards

proy30 committed:
code structure adjustment

proy30 committed:
Added space charge documentation

proy30 committed:
Added space charge documentation

proy30 committed:
Added CSR configuration

proy30 committed:
Rename 'CSR Configuration' to 'CSR' for UI and code

folder name is still called 'csrConfiguration'

proy30 committed:
Connect csr_bins to simulation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Updated UI look and minor code structure

proy30 committed:
Add initial validation for Space Charge

proy30 committed:
Add initial validation for Space Charge

proy30 committed:
update simulation code & change Vcombobox to vSelect

proy30 committed:
update simulation code & change Vcombobox to vSelect

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Updated default values

ax3l commented on src/python/impactx/dashboard/Input/space_charge_configuration/spaceChargeFunctions.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class SpaceChargeFunctions:
+    def validate_prob_relative_fields(index, prob_relative_value):
```
Please add `self` as first argument :)

ax3l commented on src/python/impactx/dashboard/Input/generalFunctions.py:
```diff
@@ -36,6 +36,10 @@ def documentation(section_name):
             url = "https://impactx.readthedocs.io/en/latest/usage/python.html#initial-beam-distributions"
         elif section_name == "pythonParameters":
             url = "https://impactx.readthedocs.io/en/latest/usage/python.html#general"
+        elif section_name == "space_charge_documentation":
+            url = "https://impactx.readthedocs.io/en/latest/usage/parameters.html#space-charge"
```
Better style: You could make this block and access a dictionary.

proy30 commented:
UI as of 9/4/24 - 6:25pm PST:

![chrome_dDj4cC6PK7](https://github.com/user-attachments/assets/751896ff-0baf-4ad1-8264-7b954fde7ae0)

ax3l commented on src/python/impactx/dashboard/Input/space_charge_configuration/spaceChargeFunctions.py:
```diff
@@ -0,0 +1,36 @@
+        """
+        error_message = ""
+
+        try:
+            prob_relative_value = float(prob_relative_value)
+            poisson_solver = state.poisson_solver
+
+            if poisson_solver == "multigrid":
```
Validation looks good.

ax3l commented on src/python/impactx/dashboard/Input/space_charge_configuration/spaceChargeMain.py:
```diff
@@ -0,0 +1,203 @@
+
+state.dynamic_size = False
+state.max_level = 3
+state.n_cell = [0.0, 0.0, 0.0]
```
`n_cell` are _integers_.

Validation: they MUST be `>1`  if space charge (any solver) is on.

A sensible, coarse default is:
```suggestion
state.n_cell = [32, 32, 32]
```

ax3l commented on src/python/impactx/dashboard/Input/space_charge_configuration/spaceChargeMain.py:
```diff
@@ -0,0 +1,203 @@
+state.n_cell = [0.0, 0.0, 0.0]
+state.prob_relative = []
+state.particle_shape = 2
+state.poisson_solver = "multigrid"
```
better default:
```suggestion
state.poisson_solver = "fft"

```

ax3l commented:
Thanks for the update screenshot, very nice.

Both space charge and CSR need a general on/off button as well to turn it on/off. If off, input validation should be deactivated and no options be added from it.

proy30 committed:
Format update

proy30 committed:
Minor updates

proy30 commented:
Accidently pulled in code from unit testing.. need to remove first 18 commits as they are not related to this PR.

proy30 committed:
Revert "Added IDs to UI components"

This reverts commit 44566659bf255f582f756a21cf022c42163eeda3.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Removed duiplicated UI cards

proy30 committed:
Formatting change and remove duplicate code

proy30 commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -40,8 +40,22 @@ def run_simulation():
     sim = ImpactX()
 
+    # space charge selections
+    if state.space_charge:
+        sim.max_level = state.max_level
+        sim.n_cell = state.n_cell
+        sim.particle_shape = state.particle_shape
+        sim.poisson_solver = state.poisson_solver
+        sim.space_charge = state.space_charge
+        sim.dynamic_size = state.dynamic_size
+        sim.prob_relative = state.prob_relative
+
+    # csr
+    if state.csr:
+        sim.csr = state.csr
+        sim.csr_bins = state.csr_bins
+
```
Currently, `particle_shape` is only displayed when `space_charge` is set to True. However, I see that `particle_shape` is also being used in `run_chicane_csr.py`. Should we consider having `particle_shape` displayed for CSR as well?

proy30 committed:
Update default n_cell

proy30 committed:
particle_shape placed in CSR

proy30 committed:
Added blocking_factor to UI, no functionality

proy30 committed:
Added blocking_factor defaults

proy30 committed:
Added blocking_factor functionality

last step is to connect to simulation

proy30 committed:
Update default values for space charge

proy30 committed:
Updated prob_relative validation

prob_relative values are required to decrease with increasing index

proy30 committed:
Connected blocking_factor to simulation

proy30 committed:
Add multigrid-specific numerical options

No validation yet and not yet connected to simulation

proy30 committed:
Adjust UI formatting

proy30 committed:
Update export_template for space charge and csr

proy30 committed:
Updated default for blocking_factor xyz

proy30 committed:
Fixed simulation.py after merge

Space charge/csr simulation content did not merge in correctly originally, this fixes it.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Merge remote-tracking branch 'mainline/development' into dashboard/space_charge_configuration

proy30 committed:
Merge branch 'dashboard/space_charge_configuration' of github.com:proy30/impactx into dashboard/space_charge_configuration

proy30 committed:
Update validation for n_cell and blocking_factor

proy30 committed:
Updated ui format

proy30 committed:
Add csr validation

csr_bins int and positive number

proy30 committed:
Update simulation station function

to ensure that users cannot run a simulation when inputs have errors

proy30 committed:
Merge branch 'development' into dashboard/space_charge_configuration

EZoni commented on src/python/impactx/dashboard/simulation.py:
```diff
@@ -1,81 +1,99 @@
+        sim.space_charge = state.space_charge
+        sim.dynamic_size = state.dynamic_size
+        sim.prob_relative = state.prob_relative
+        sim.blocking_factor_x = [state.blocking_factor_x]
+        sim.blocking_factor_y = [state.blocking_factor_y]
+        sim.blocking_factor_z = [state.blocking_factor_z]
```
I can't remember how this works on the AMReX end, so just to double check, do these have to be single-element lists or can they be simply integers?

cemitch99 replied:
Looking at [this](https://amrex-codes.github.io/amrex/docs_html/RuntimeParameters.html?highlight=blocking+factor#amr.blocking_factor_x), it looks like this can be a single integer.  Indeed, we use a single integer in this [example](https://github.com/ECP-WarpX/impactx/blob/development/examples/expanding_beam/input_expanding_fft.in).  Unfortunately the ImpactX documentation does not appear to address the blocking factor, so this needs to be updated.

EZoni commented on src/python/impactx/dashboard/Toolbar/exportTemplate.py:
Not sure if this is a git display issue, but the indentation of the function body for `build_space_charge_or_csr` seems off. Could you double check?

EZoni commented on src/python/impactx/dashboard/Input/space_charge_configuration/spaceChargeMain.py:
```diff
@@ -0,0 +1,328 @@
+                            with vuetify.VCol(cols=6, classes="py-0"):
+                                vuetify.VTextField(
+                                    label="MLMG Relative Tolerance",
+                                    v_model=("mlmg_relative_tolerance",),
```
In one discussion in the Trame repo, I was told that one could use `v_model_number` as well (on top of `type="number"`) and that it could render numbers better when it comes to formatting.

I did try it on my end without noticing any significant improvement, but maybe you could try it as well.

For reference, here's the comment where this was mentioned: https://github.com/Kitware/trame/discussions/599#discussioncomment-10984724.

EZoni commented on src/python/impactx/dashboard/Input/space_charge_configuration/spaceChargeMain.py:
```diff
@@ -0,0 +1,328 @@
+def populate_prob_relative_fields(max_level):
+    num_prob_relative_fields = int(max_level) + 1
+
+    if state.poisson_solver == "fft":
+        state.prob_relative = [1.1] + [0.0] * (num_prob_relative_fields - 1)
+    elif state.poisson_solver == "multigrid":
+        state.prob_relative = [3.1] + [0.0] * (num_prob_relative_fields - 1)
+    else:
+        state.prob_relative = [0.0] * num_prob_relative_fields
```
Question for both you and @cemitch99, should we add a comment about the numbers that appear here (e.g., `1.1`, `3.1`, etc.)? I'm not too familiar with the context here, so I'm not sure if the meaning of these numbers should be clarified.

cemitch99 replied:
Those numbers in `prob_relative` specify the size of the grid relative to the size of the beam, and are documented [here](https://impactx.readthedocs.io/en/latest/usage/parameters.html#collective-effects) and [here](https://impactx.readthedocs.io/en/latest/usage/python.html#collective-effects-overall-simulation-parameters).  Maybe some of the input parameters should include links to the documentation?

cemitch99 reviewed (approved):
This looks good to me.  We can discuss a little when we chat today.

proy30 committed:
connect multigrid advanced settings to simulation

proy30 committed:
Merge remote-tracking branch 'mainline/development' into dashboard/space_charge_configuration

proy30 committed:
add space

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
